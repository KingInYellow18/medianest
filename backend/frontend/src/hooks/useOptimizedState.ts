import { useCallback, useMemo, useRef, useState } from 'react';

// Context7 Pattern - Branded types for state management
type StateId = string & { readonly __brand: 'StateId' };
type StateVersion = number & { readonly __brand: 'StateVersion' };

// Context7 Pattern - Result type for state operations
type StateResult<T, E = Error> =
  | { readonly success: true; readonly data: T; readonly version: StateVersion }
  | { readonly success: false; readonly error: E; readonly version: StateVersion };

// Context7 Pattern - Optimized state hook with memoization
export function useOptimizedState<T>(
  initialState: T | (() => T),
  stateId?: string
): [T, (newState: T | ((prev: T) => T)) => void, { version: StateVersion; reset: () => void }] {
  const id = useMemo(() => (stateId || Date.now().toString()) as StateId, [stateId]);
  const versionRef = useRef<StateVersion>(0 as StateVersion);
  const [state, setState] = useState<T>(initialState);

  // Context7 Pattern - Memoized state updater
  const updateState = useCallback((newState: T | ((prev: T) => T)) => {
    setState((prevState) => {
      versionRef.current = (versionRef.current + 1) as StateVersion;
      return typeof newState === 'function' ? (newState as (prev: T) => T)(prevState) : newState;
    });
  }, []);

  // Context7 Pattern - Memoized reset function
  const reset = useCallback(() => {
    versionRef.current = 0 as StateVersion;
    setState(initialState);
  }, [initialState]);

  // Context7 Pattern - Memoized metadata object
  const metadata = useMemo(
    () => ({
      version: versionRef.current,
      reset,
    }),
    [versionRef.current, reset]
  );

  return [state, updateState, metadata];
}

// Context7 Pattern - Type-safe async state hook
export function useAsyncState<T, E = Error>(
  asyncFn?: () => Promise<T>
): {
  data: T | null;
  loading: boolean;
  error: E | null;
  execute: (fn?: () => Promise<T>) => Promise<StateResult<T, E>>;
  reset: () => void;
} {
  const [state, setState] = useOptimizedState<{
    data: T | null;
    loading: boolean;
    error: E | null;
    version: StateVersion;
  }>({
    data: null,
    loading: false,
    error: null,
    version: 0 as StateVersion,
  });

  // Context7 Pattern - Memoized execution function
  const execute = useCallback(
    async (fn?: () => Promise<T>): Promise<StateResult<T, E>> => {
      const executeFn = fn || asyncFn;
      if (!executeFn) {
        const error = new Error('No async function provided') as E;
        setState((prev) => ({
          ...prev,
          error,
          version: (prev.version + 1) as StateVersion,
        }));
        return { success: false, error, version: state.version };
      }

      setState((prev) => ({
        ...prev,
        loading: true,
        error: null,
        version: (prev.version + 1) as StateVersion,
      }));

      try {
        const data = await executeFn();
        const newVersion = (state.version + 1) as StateVersion;
        setState((prev) => ({
          data,
          loading: false,
          error: null,
          version: newVersion,
        }));
        return { success: true, data, version: newVersion };
      } catch (error) {
        const typedError = error as E;
        const newVersion = (state.version + 1) as StateVersion;
        setState((prev) => ({
          ...prev,
          loading: false,
          error: typedError,
          version: newVersion,
        }));
        return { success: false, error: typedError, version: newVersion };
      }
    },
    [asyncFn, state.version, setState]
  );

  // Context7 Pattern - Memoized reset function
  const reset = useCallback(() => {
    setState({
      data: null,
      loading: false,
      error: null,
      version: 0 as StateVersion,
    });
  }, [setState]);

  return {
    data: state.data,
    loading: state.loading,
    error: state.error,
    execute,
    reset,
  };
}

// Context7 Pattern - Optimized debounced state hook
export function useDebouncedState<T>(
  initialValue: T,
  delay: number = 300
): [T, T, (value: T) => void] {
  const [immediateValue, setImmediateValue] = useState<T>(initialValue);
  const [debouncedValue, setDebouncedValue] = useState<T>(initialValue);
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);

  // Context7 Pattern - Memoized debounced setter
  const setValue = useCallback(
    (value: T) => {
      setImmediateValue(value);

      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }

      timeoutRef.current = setTimeout(() => {
        setDebouncedValue(value);
      }, delay);
    },
    [delay]
  );

  // Cleanup on unmount
  useMemo(() => {
    return () => {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current);
      }
    };
  }, []);

  return [immediateValue, debouncedValue, setValue];
}
