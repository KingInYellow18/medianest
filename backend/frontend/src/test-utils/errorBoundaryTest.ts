/**
 * Utility for testing React Error Boundaries with Vitest
 *
 * This provides a clean way to test error boundaries by temporarily
 * suppressing React's error logging during error boundary tests.
 */

// Store original console.error function
const originalError = console.error;

export const suppressReactErrorLogs = (callback: () => void) => {
  // Temporarily suppress console.error to prevent React error logs during tests
  console.error = () => {};

  try {
    callback();
  } finally {
    // Restore original console.error
    console.error = originalError;
  }
};

export const withErrorBoundarySuppression = (testFn: () => void | Promise<void>) => {
  return () => {
    console.error = () => {};
    try {
      return testFn();
    } finally {
      console.error = originalError;
    }
  };
};
