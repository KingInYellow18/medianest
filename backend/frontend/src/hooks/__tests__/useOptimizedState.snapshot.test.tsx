/**
 * Snapshot Tests for useOptimizedState Hook
 * Tests visual consistency of components using optimized state management
 */

import React, { useCallback } from 'react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, fireEvent, act } from '../../test-utils/render';
import { useOptimizedState, useAsyncState, useDebouncedState } from '../useOptimizedState';

// Test component for useOptimizedState
const StateComponent = ({ initialValue }: { initialValue: any }) => {
  const [state, setState, meta] = useOptimizedState(initialValue);
  
  return (
    <div data-testid="state-component">
      <div data-testid="state-value">{JSON.stringify(state)}</div>
      <div data-testid="state-version">{meta.version}</div>
      <div className="state-controls">
        <button 
          data-testid="increment-btn"
          onClick={() => setState((prev: number) => prev + 1)}
        >
          Increment
        </button>
        <button 
          data-testid="set-object-btn"
          onClick={() => setState({ name: 'Test', count: 42 })}
        >
          Set Object
        </button>
        <button 
          data-testid="reset-btn"
          onClick={meta.reset}
        >
          Reset
        </button>
      </div>
    </div>
  );
};

// Test component for useAsyncState
const AsyncStateComponent = () => {
  const mockAsyncFn = useCallback(async () => {
    await new Promise(resolve => setTimeout(resolve, 100));
    return { data: 'async result', timestamp: new Date().toISOString() };
  }, []);

  const { data, loading, error, execute, reset } = useAsyncState(mockAsyncFn);
  
  return (
    <div data-testid="async-state-component">
      <div data-testid="async-data">{data ? JSON.stringify(data) : 'No data'}</div>
      <div data-testid="async-loading">{loading ? 'Loading...' : 'Not loading'}</div>
      <div data-testid="async-error">{error ? error.message : 'No error'}</div>
      <div className="async-controls">
        <button 
          data-testid="execute-btn"
          onClick={() => execute()}
        >
          Execute
        </button>
        <button 
          data-testid="execute-failing-btn"
          onClick={() => execute(async () => {
            throw new Error('Async operation failed');
          })}
        >
          Execute Failing
        </button>
        <button 
          data-testid="reset-async-btn"
          onClick={reset}
        >
          Reset
        </button>
      </div>
    </div>
  );
};

// Test component for useDebouncedState
const DebouncedStateComponent = ({ delay = 300 }: { delay?: number }) => {
  const [immediate, debounced, setValue] = useDebouncedState('initial', delay);
  
  return (
    <div data-testid="debounced-state-component">
      <div data-testid="immediate-value">{immediate}</div>
      <div data-testid="debounced-value">{debounced}</div>
      <div className="debounced-controls">
        <input 
          data-testid="debounced-input"
          value={immediate}
          onChange={(e) => setValue(e.target.value)}
          placeholder="Type to test debouncing"
        />
        <button 
          data-testid="set-quick-btn"
          onClick={() => setValue('quick change')}
        >
          Quick Change
        </button>
        <button 
          data-testid="set-long-text-btn"
          onClick={() => setValue('This is a very long text value that tests how debounced state handles longer strings')}
        >
          Set Long Text
        </button>
      </div>
    </div>
  );
};

// Combined component for complex state scenarios
const CombinedStateComponent = () => {
  const [counter, setCounter] = useOptimizedState(0);
  const [text, debouncedText, setText] = useDebouncedState('', 200);
  const { data: asyncData, loading, execute } = useAsyncState();
  
  return (
    <div data-testid="combined-state-component">
      <div className="counter-section">
        <h3>Counter State</h3>
        <div data-testid="counter-value">{counter}</div>
        <button onClick={() => setCounter(prev => prev + 1)}>Increment</button>
      </div>
      
      <div className="text-section">
        <h3>Debounced Text</h3>
        <div data-testid="immediate-text">{text}</div>
        <div data-testid="debounced-text">{debouncedText}</div>
        <input 
          value={text}
          onChange={(e) => setText(e.target.value)}
          placeholder="Type here..."
        />
      </div>
      
      <div className="async-section">
        <h3>Async Data</h3>
        <div data-testid="combined-async-data">
          {loading ? 'Loading...' : (asyncData ? JSON.stringify(asyncData) : 'No data')}
        </div>
        <button 
          onClick={() => execute(async () => ({ result: 'Combined async result' }))}
        >
          Load Data
        </button>
      </div>
    </div>
  );
};

describe('useOptimizedState Hook Snapshot Tests', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe('useOptimizedState Snapshots', () => {
    it('should match snapshot with initial number value', () => {
      const { container } = render(<StateComponent initialValue={0} />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-initial-number');
    });

    it('should match snapshot with initial string value', () => {
      const { container } = render(<StateComponent initialValue="initial string" />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-initial-string');
    });

    it('should match snapshot with initial object value', () => {
      const initialObject = { name: 'Test Object', value: 42, nested: { deep: true } };
      const { container } = render(<StateComponent initialValue={initialObject} />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-initial-object');
    });

    it('should match snapshot with initial array value', () => {
      const initialArray = [1, 2, 3, 'four', { five: 5 }];
      const { container } = render(<StateComponent initialValue={initialArray} />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-initial-array');
    });

    it('should match snapshot after state increment', () => {
      const { container, getByTestId } = render(<StateComponent initialValue={5} />);
      
      const incrementBtn = getByTestId('increment-btn');
      act(() => {
        fireEvent.click(incrementBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-after-increment');
    });

    it('should match snapshot after multiple increments', () => {
      const { container, getByTestId } = render(<StateComponent initialValue={0} />);
      
      const incrementBtn = getByTestId('increment-btn');
      act(() => {
        fireEvent.click(incrementBtn);
        fireEvent.click(incrementBtn);
        fireEvent.click(incrementBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-multiple-increments');
    });

    it('should match snapshot after setting object', () => {
      const { container, getByTestId } = render(<StateComponent initialValue={0} />);
      
      const setObjectBtn = getByTestId('set-object-btn');
      act(() => {
        fireEvent.click(setObjectBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-after-set-object');
    });

    it('should match snapshot after reset', () => {
      const { container, getByTestId } = render(<StateComponent initialValue={10} />);
      
      const incrementBtn = getByTestId('increment-btn');
      const resetBtn = getByTestId('reset-btn');
      
      act(() => {
        fireEvent.click(incrementBtn);
        fireEvent.click(incrementBtn);
        fireEvent.click(resetBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-after-reset');
    });
  });

  describe('useAsyncState Snapshots', () => {
    it('should match snapshot in initial state', () => {
      const { container } = render(<AsyncStateComponent />);
      
      expect(container.firstChild).toMatchSnapshot('async-state-initial');
    });

    it('should match snapshot during loading', () => {
      const { container, getByTestId } = render(<AsyncStateComponent />);
      
      const executeBtn = getByTestId('execute-btn');
      act(() => {
        fireEvent.click(executeBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('async-state-loading');
    });

    it('should match snapshot after successful execution', async () => {
      const { container, getByTestId } = render(<AsyncStateComponent />);
      
      const executeBtn = getByTestId('execute-btn');
      act(() => {
        fireEvent.click(executeBtn);
      });
      
      // Fast forward timers to complete the async operation
      act(() => {
        vi.advanceTimersByTime(200);
      });
      
      expect(container.firstChild).toMatchSnapshot('async-state-success');
    });

    it('should match snapshot with error state', () => {
      const { container, getByTestId } = render(<AsyncStateComponent />);
      
      const executeFailingBtn = getByTestId('execute-failing-btn');
      act(() => {
        fireEvent.click(executeFailingBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('async-state-error');
    });

    it('should match snapshot after reset from error', () => {
      const { container, getByTestId } = render(<AsyncStateComponent />);
      
      const executeFailingBtn = getByTestId('execute-failing-btn');
      const resetBtn = getByTestId('reset-async-btn');
      
      act(() => {
        fireEvent.click(executeFailingBtn);
        fireEvent.click(resetBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('async-state-after-reset');
    });
  });

  describe('useDebouncedState Snapshots', () => {
    it('should match snapshot with initial debounced state', () => {
      const { container } = render(<DebouncedStateComponent />);
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-initial');
    });

    it('should match snapshot with immediate value change', () => {
      const { container, getByTestId } = render(<DebouncedStateComponent />);
      
      const input = getByTestId('debounced-input');
      act(() => {
        fireEvent.change(input, { target: { value: 'immediate change' } });
      });
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-immediate-change');
    });

    it('should match snapshot after debounce delay', () => {
      const { container, getByTestId } = render(<DebouncedStateComponent delay={100} />);
      
      const input = getByTestId('debounced-input');
      act(() => {
        fireEvent.change(input, { target: { value: 'debounced value' } });
      });
      
      // Fast forward past debounce delay
      act(() => {
        vi.advanceTimersByTime(150);
      });
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-after-delay');
    });

    it('should match snapshot with rapid changes', () => {
      const { container, getByTestId } = render(<DebouncedStateComponent delay={200} />);
      
      const input = getByTestId('debounced-input');
      act(() => {
        fireEvent.change(input, { target: { value: 'first' } });
        fireEvent.change(input, { target: { value: 'second' } });
        fireEvent.change(input, { target: { value: 'final' } });
      });
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-rapid-changes');
    });

    it('should match snapshot with long text value', () => {
      const { container, getByTestId } = render(<DebouncedStateComponent />);
      
      const setLongTextBtn = getByTestId('set-long-text-btn');
      act(() => {
        fireEvent.click(setLongTextBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-long-text');
    });

    it('should match snapshot with custom delay', () => {
      const { container } = render(<DebouncedStateComponent delay={1000} />);
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-custom-delay');
    });
  });

  describe('Combined State Scenarios Snapshots', () => {
    it('should match snapshot of combined state component initial state', () => {
      const { container } = render(<CombinedStateComponent />);
      
      expect(container.firstChild).toMatchSnapshot('combined-state-initial');
    });

    it('should match snapshot with all states active', async () => {
      const { container, getByText, getByPlaceholderText } = render(<CombinedStateComponent />);
      
      // Increment counter
      const incrementBtn = getByText('Increment');
      act(() => {
        fireEvent.click(incrementBtn);
        fireEvent.click(incrementBtn);
      });
      
      // Change text
      const textInput = getByPlaceholderText('Type here...');
      act(() => {
        fireEvent.change(textInput, { target: { value: 'Testing combined state' } });
      });
      
      // Start async operation
      const loadBtn = getByText('Load Data');
      act(() => {
        fireEvent.click(loadBtn);
      });
      
      expect(container.firstChild).toMatchSnapshot('combined-state-all-active');
    });

    it('should match snapshot with complex interactions', () => {
      const { container, getByText, getByPlaceholderText } = render(<CombinedStateComponent />);
      
      // Complex sequence of interactions
      const incrementBtn = getByText('Increment');
      const textInput = getByPlaceholderText('Type here...');
      const loadBtn = getByText('Load Data');
      
      act(() => {
        // Multiple increments
        for (let i = 0; i < 5; i++) {
          fireEvent.click(incrementBtn);
        }
        
        // Text changes
        fireEvent.change(textInput, { target: { value: 'First change' } });
        fireEvent.change(textInput, { target: { value: 'Second change' } });
        fireEvent.change(textInput, { target: { value: 'Final text value' } });
        
        // Async operation
        fireEvent.click(loadBtn);
        
        // Advance timers for debounce
        vi.advanceTimersByTime(250);
      });
      
      expect(container.firstChild).toMatchSnapshot('combined-state-complex-interactions');
    });
  });

  describe('Edge Cases and Error Conditions Snapshots', () => {
    it('should match snapshot with null initial value', () => {
      const { container } = render(<StateComponent initialValue={null} />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-null-initial');
    });

    it('should match snapshot with undefined initial value', () => {
      const { container } = render(<StateComponent initialValue={undefined} />);
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-undefined-initial');
    });

    it('should match snapshot with empty string debounced state', () => {
      const { container } = render(<DebouncedStateComponent />);
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-empty-string');
    });

    it('should match snapshot with zero delay debounced state', () => {
      const { container, getByTestId } = render(<DebouncedStateComponent delay={0} />);
      
      const input = getByTestId('debounced-input');
      act(() => {
        fireEvent.change(input, { target: { value: 'zero delay test' } });
        vi.advanceTimersByTime(10);
      });
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-zero-delay');
    });

    it('should match snapshot with very large delay', () => {
      const { container } = render(<DebouncedStateComponent delay={10000} />);
      
      expect(container.firstChild).toMatchSnapshot('debounced-state-large-delay');
    });
  });

  describe('Performance Optimization Snapshots', () => {
    it('should match snapshot with rapid state updates', () => {
      const { container, getByTestId } = render(<StateComponent initialValue={0} />);
      
      const incrementBtn = getByTestId('increment-btn');
      act(() => {
        // Simulate rapid clicking
        for (let i = 0; i < 10; i++) {
          fireEvent.click(incrementBtn);
        }
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-rapid-updates');
    });

    it('should match snapshot with state id optimization', () => {
      const StateWithIdComponent = () => {
        const [state, setState] = useOptimizedState(42, 'unique-state-id');
        return (
          <div data-testid="state-with-id">
            <div data-testid="state-id-value">{state}</div>
            <button onClick={() => setState(prev => prev * 2)}>Double</button>
          </div>
        );
      };

      const { container, getByText } = render(<StateWithIdComponent />);
      
      act(() => {
        fireEvent.click(getByText('Double'));
      });
      
      expect(container.firstChild).toMatchSnapshot('optimized-state-with-id');
    });
  });
});