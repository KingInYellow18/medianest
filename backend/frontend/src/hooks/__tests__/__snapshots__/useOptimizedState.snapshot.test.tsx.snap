// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useOptimizedState Hook Snapshot Tests > Combined State Scenarios Snapshots > should match snapshot with complex interactions > combined-state-initial 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="combined-state-component"
  >
    <div
      class="counter-section"
    >
      <h3>
        Counter State
      </h3>
      <div
        data-testid="counter-value"
      >
        0
      </div>
      <button>
        Increment
      </button>
    </div>
    <div
      class="text-section"
    >
      <h3>
        Debounced Text
      </h3>
      <div
        data-testid="immediate-text"
      />
      <div
        data-testid="debounced-text"
      />
      <input
        placeholder="Type here..."
        value=""
      />
    </div>
    <div
      class="async-section"
    >
      <h3>
        Async Data
      </h3>
      <div
        data-testid="combined-async-data"
      >
        No data
      </div>
      <button>
        Load Data
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Edge Cases and Error Conditions Snapshots > should match snapshot with very large delay > debounced-state-large-delay 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="debounced-state-component"
  >
    <div
      data-testid="immediate-value"
    >
      initial
    </div>
    <div
      data-testid="debounced-value"
    >
      initial
    </div>
    <div
      class="debounced-controls"
    >
      <input
        data-testid="debounced-input"
        placeholder="Type to test debouncing"
        value="initial"
      />
      <button
        data-testid="set-quick-btn"
      >
        Quick Change
      </button>
      <button
        data-testid="set-long-text-btn"
      >
        Set Long Text
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Edge Cases and Error Conditions Snapshots > should match snapshot with zero delay debounced state > debounced-state-empty-string 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="debounced-state-component"
  >
    <div
      data-testid="immediate-value"
    >
      initial
    </div>
    <div
      data-testid="debounced-value"
    >
      initial
    </div>
    <div
      class="debounced-controls"
    >
      <input
        data-testid="debounced-input"
        placeholder="Type to test debouncing"
        value="initial"
      />
      <button
        data-testid="set-quick-btn"
      >
        Quick Change
      </button>
      <button
        data-testid="set-long-text-btn"
      >
        Set Long Text
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Edge Cases and Error Conditions Snapshots > should match snapshot with zero delay debounced state > optimized-state-null-initial 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      null
    </div>
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Edge Cases and Error Conditions Snapshots > should match snapshot with zero delay debounced state > optimized-state-undefined-initial 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    />
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Performance Optimization Snapshots > should match snapshot with state id optimization > optimized-state-rapid-updates 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      10
    </div>
    <div
      data-testid="state-version"
    >
      10
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > Performance Optimization Snapshots > should match snapshot with state id optimization > optimized-state-with-id 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-with-id"
  >
    <div
      data-testid="state-id-value"
    >
      84
    </div>
    <button>
      Double
    </button>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useAsyncState Snapshots > should match snapshot after reset from error > async-state-after-reset 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="async-state-component"
  >
    <div
      data-testid="async-data"
    >
      No data
    </div>
    <div
      data-testid="async-loading"
    >
      Not loading
    </div>
    <div
      data-testid="async-error"
    >
      No error
    </div>
    <div
      class="async-controls"
    >
      <button
        data-testid="execute-btn"
      >
        Execute
      </button>
      <button
        data-testid="execute-failing-btn"
      >
        Execute Failing
      </button>
      <button
        data-testid="reset-async-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useAsyncState Snapshots > should match snapshot with error state > async-state-initial 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="async-state-component"
  >
    <div
      data-testid="async-data"
    >
      No data
    </div>
    <div
      data-testid="async-loading"
    >
      Not loading
    </div>
    <div
      data-testid="async-error"
    >
      No error
    </div>
    <div
      class="async-controls"
    >
      <button
        data-testid="execute-btn"
      >
        Execute
      </button>
      <button
        data-testid="execute-failing-btn"
      >
        Execute Failing
      </button>
      <button
        data-testid="reset-async-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useDebouncedState Snapshots > should match snapshot with custom delay > debounced-state-custom-delay 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="debounced-state-component"
  >
    <div
      data-testid="immediate-value"
    >
      initial
    </div>
    <div
      data-testid="debounced-value"
    >
      initial
    </div>
    <div
      class="debounced-controls"
    >
      <input
        data-testid="debounced-input"
        placeholder="Type to test debouncing"
        value="initial"
      />
      <button
        data-testid="set-quick-btn"
      >
        Quick Change
      </button>
      <button
        data-testid="set-long-text-btn"
      >
        Set Long Text
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useDebouncedState Snapshots > should match snapshot with custom delay > debounced-state-long-text 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="debounced-state-component"
  >
    <div
      data-testid="immediate-value"
    >
      This is a very long text value that tests how debounced state handles longer strings
    </div>
    <div
      data-testid="debounced-value"
    >
      initial
    </div>
    <div
      class="debounced-controls"
    >
      <input
        data-testid="debounced-input"
        placeholder="Type to test debouncing"
        value="This is a very long text value that tests how debounced state handles longer strings"
      />
      <button
        data-testid="set-quick-btn"
      >
        Quick Change
      </button>
      <button
        data-testid="set-long-text-btn"
      >
        Set Long Text
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useDebouncedState Snapshots > should match snapshot with rapid changes > debounced-state-initial 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="debounced-state-component"
  >
    <div
      data-testid="immediate-value"
    >
      initial
    </div>
    <div
      data-testid="debounced-value"
    >
      initial
    </div>
    <div
      class="debounced-controls"
    >
      <input
        data-testid="debounced-input"
        placeholder="Type to test debouncing"
        value="initial"
      />
      <button
        data-testid="set-quick-btn"
      >
        Quick Change
      </button>
      <button
        data-testid="set-long-text-btn"
      >
        Set Long Text
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useOptimizedState Snapshots > should match snapshot after reset > optimized-state-after-increment 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      6
    </div>
    <div
      data-testid="state-version"
    >
      1
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useOptimizedState Snapshots > should match snapshot with initial array value > optimized-state-initial-array 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      [1,2,3,"four",{"five":5}]
    </div>
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useOptimizedState Snapshots > should match snapshot with initial array value > optimized-state-initial-number 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      0
    </div>
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useOptimizedState Snapshots > should match snapshot with initial array value > optimized-state-initial-object 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      {"name":"Test Object","value":42,"nested":{"deep":true}}
    </div>
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;

exports[`useOptimizedState Hook Snapshot Tests > useOptimizedState Snapshots > should match snapshot with initial array value > optimized-state-initial-string 1`] = `
<div
  data-testid="test-wrapper"
>
  <div
    data-testid="state-component"
  >
    <div
      data-testid="state-value"
    >
      "initial string"
    </div>
    <div
      data-testid="state-version"
    >
      0
    </div>
    <div
      class="state-controls"
    >
      <button
        data-testid="increment-btn"
      >
        Increment
      </button>
      <button
        data-testid="set-object-btn"
      >
        Set Object
      </button>
      <button
        data-testid="reset-btn"
      >
        Reset
      </button>
    </div>
  </div>
</div>
`;
